### ISSUE TITLE
"Add page" list incorrect after changing allowed_children

### ISSUE THREAD

**Issue by spekulatius** [silverstripe/silverstripe-cms#2450]
Hello SilverStripers,

I've noticed something odd which might be a bug. On a client website I've implemented an extension tweaking the allowed_children to reduce possible choices. The new list is stored in the configuration of the page. When I open the "Add Page" the options are disabled as expected. Shortly after, the correctly disabled options getting enabled. Here is a short video (as a gif for GitHub) of the effect:

![ezgif-4-b9196c70b92f](https://user-images.githubusercontent.com/8433587/60008220-de974c00-969d-11e9-889b-8d615c06e6fd.gif)

Just before the incorrect reset of the disabled page types happens, a call to `/admin/pages/add/childfilter?ParentID=14` is returned with the correct values:

![image](https://user-images.githubusercontent.com/8433587/60008950-2ff40b00-969f-11e9-8f0d-b10d089bc965.png)

I'm unsure if I maybe have implemented the limitation incorrect (I'm writing to the config) or this is a bug. How can I debug this more?

Anyone got an idea on this? Maybe again @robbieaverill or @NightJar ?

Cheers,
Peter

**Comment by NightJar** [comment]
Hi @spekulatius, I'm not actually a core committer sorry - the direct tagging is flattering, but misdirected :)

I will make some stabs in the dark for you though.

If I didn't know better I'd guess that there's a state machine that's being updated in the background, such as `redux` for the `React` based interface, which is checking it's local cache on load.

I (think I) do know better though, and this is (more than likely) all still `jQuery.entwine.js` - I still maintain that the local cache update is _along_ the right lines, but I'd suggest it's a plain pjax request or whatnot that entwine is using `onadd` or suchlike - the second request obviously isn't your first one, and therefore does not have the config alterations that you performed in it.

Thus you get an updated list of "vanilla" page type selection options.

Good luck debugging :)

**Comment by spekulatius** [comment]
Hello @NightJar,

haha okay, I've just been taking how responded last time :)

yeah, something isn't updating correctly. Just don't have the insights to finding the issue really. I've just plainly expected if the allowed_children excludes a Page type it would update the listing with it. Digging into the code for days doesn't sound like a good approach. Any more ideas where to look exactly are highly appreciated. 

Peter

**Comment by NightJar** [comment]
>Digging into the code for days doesn't sound like a good approach.

Fair enough 😄 
Try here: https://github.com/silverstripe/silverstripe-cms/blob/4/client/src/legacy/CMSMain.AddForm.js#L67 :)

**Comment by chillu** [comment]
@spekulatius Which version of the module is this on? Can you do a `composer info`? Somewhat related: https://github.com/silverstripe/silverstripe-cms/issues/2207

**Comment by spekulatius** [comment]
Hello @NightJar and @chillu 

I'll have a look and see if I can identify the issue @NightJar. 

Ingo, the CMS and Framework are 4.4.1. I can provide a complete list if it helps :)

Peter

**Comment by axllent** [comment]
It appears to me that `$allowed_children` isn't working visually at all in 4.4.[0-1] - right mouse clicking on a page type that has `private static $allowed_children = "none";`, or ones that have a set array still present all the page types in `Add new page here`, and also appear when you go try "Add page" via the button.

**Comment by spekulatius** [comment]
> It appears to me that `$allowed_children` isn't working visually at all in 4.4.[0-1] - right mouse clicking on a page type that has `private static $allowed_children = "none";`, or ones that have a set array still present all the page types in `Add new page here`, and also appear when you go try "Add page" via the button.

Yes, that's what it looks like to me as well. It also allows me to actually create the page.

**Comment by axllent** [comment]
@spekulatius I managed to solve this issue. It was being caused by `canCreate()` permissions in my `Page::class`. Previous versions of SIlverStripe would use `canCreate()` and if `true` then also check `$allowed_children`, but it seems that now if `canCreate()` returns `true` then this overrides anything in `$allowed_children` (with regards to the dropdown at least). 

This may point you in the right direction....

**Comment by ToddHossack** [comment]
I've experienced the same here with version 4.4.3. The canCreate method overrides the allowed children. Further, after selecting the page type and clicking the "Create" button, the UI does not proceed to the new page form.
